sklearn/preprocessing/_encoders.py

```diff
@@ -398,9 +409,21 @@ def transform(self, X):
             # categories that occur after them to avoid an empty column.
 
             keep_cells = X_int != to_drop
+
```

Comment:
For this `if self.drop is not None` block, this may be cleaner:

```py
        if self.drop is not None:
            to_drop = self.drop_idx_.reshape(1, -1).copy()

            keep_cells = X_int != to_drop
            n_values = []
            for i, cats in enumerate(self.categories_):
                n_cats = len(cats)
                if to_drop[0, i] is None:
                    # set to cardinality to not drop from X_int
                    to_drop[0, i] = n_cats
                    n_values.append(n_cats)
                else:  # dropped
                    n_values.append(n_cats - 1)

            X_int[X_int > to_drop] -= 1
            X_mask &= keep_cells
        else:
            n_values = [len(cats) for cats in self.categories_]

```